File: 144-binary-tree-preorder-traversal/144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> preorder = new ArrayList<>();
        if(root == null) return preorder;
        TreeNode current = root;
        while(current != null) {
            if(current.left != null) {
                TreeNode leftSubTree = current.left;
                while(leftSubTree.right != null && leftSubTree.right != current)
                    leftSubTree = leftSubTree.right;
                if(leftSubTree.right == null) {
                    leftSubTree.right = current;
                    preorder.add(current.val);
                    current = current.left;
                } else {
                    leftSubTree.right = null;
                    current = current.right;
                }
            } else {
                preorder.add(current.val);
                current = current.right;
            }
        }
        return preorder;
    }
}

File: 144-binary-tree-preorder-traversal/144-binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        preorder = []
        if not root: return preorder
        current = root
        while current:
            if current.left:
                leftSubTree = current.left
                while leftSubTree.right and leftSubTree.right != current:
                    leftSubTree = leftSubTree.right
                if not leftSubTree.right:
                    leftSubTree.right = current
                    preorder.append(current.val)
                    current = current.left
                else:
                    leftSubTree.right = None
                    current = current.right
            else:
                preorder.append(current.val)
                current = current.right
        return preorder